espnet2/train/abs_espnet_model.py

```diff
@@ -38,3 +38,15 @@ def forward(
     @abstractmethod
     def collect_feats(self, **batch: torch.Tensor) -> Dict[str, torch.Tensor]:
         raise NotImplementedError
+
```

Comment:
@ftshijt and @pyf98
Do we need this, or is there an alternative approach?

Reply:
I feel we may not want to add training-related functions to the model definition. These two concepts should be more separated, otherwise we cannot transfer the model to other trainers.

An alternative approach is to use PyTorch Lightning trainer, which provides various hooks to inject those special functions. Since this is a new task, we don't need to care about compatibility with old trainer. This is a great opportunity to test the new trainer and to accumulate experience for next-gen ESPnet development.

Reply:
`training-related functions to the model definition. These two concepts should be more separated`
+1 to this. I tried to keep the names different from pytorch lightning by appending a `_` but I think creating a hook is a cleaner solution.

Reply:
Working on this, @juice500ml noticed that current implementation does not support distributed training, so please do not merge this!

Reply:
Completed this implementation. This CL now contains changes for adding user callbacks to pytorch lightning trainer.